rfd/0003-plugins-unknown-events.md

```diff
@@ -0,0 +1,352 @@
+that are deployed independently. This means that the event handler can be
+deployed with a version that is not up to date with the Auth server version.
+
+When the event handler is not up to date with the Auth server, it will not be
+able to handle new events that are not known to the event handler. This can
+cause the event handler to crash (previous behavior) or ignore the unknown events.
+
+This is problematic because the event handler is a critical component that
+should be able to handle all types of events. If the event handler is not able
+to handle new events, it can miss important events and drift from the Auth server truth.
+
+The Teleport event handler receives events from the Auth server, serializes them
+using JSON and sends them to the downstream service (e.g. Splunk, FluentD, etc).
+When the event handler is not up to date with the Auth server, the new events
+will not be known to the event handler and the protobuf raw cannot be deserialized.
+Without the deserialized event, the event handler cannot serialize the event
+using JSON and send it to the upstream service. We could send the raw protobuf
+to auth server and have the auth server serialize it to JSON but this is not
+ideal because it will add a lot of overhead to the Auth server and will add
+a lot of latency to the event handler.
+
+Currently, the event handler ignores unknown events.
+
+## Proposal
+
+The proposal consists of having Auth server serve unstructured events when the
+client requests those endpoints. This will allow the event handler to receive
+the events with a JSON-like structure and send them to the upstream service without having to
+deserialize the Protobuf and serialize it again to JSON.
+
+To achieve this, we will add a new Protobuf service - `AuditLogService` - to Auth server that will return
+the events in an unstructured format. The event handler will use these endpoints to retrieve the events
+and session events and convert them into plain JSON before pushing upstream.
+
+The new service will be:
+
+```protobuf
+
+syntax = "proto3";
+
+package teleport.auditlog.v1;
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/auditlog/v1;auditlogv1";
+
+// AuditLogService provides methods to access audit log.
+service AuditLogService {
+  // StreamUnstructuredSessionEvents streams audit events from a given session recording in an unstructured format.
+  // This endpoint is used by the event handler to retrieve the session events as JSON.
+  rpc StreamUnstructuredSessionEvents(StreamUnstructuredSessionEventsRequest) returns (stream EventUnstructured);
+  // GetUnstructuredEvents gets events from the audit log in an unstructured format.
+  // This endpoint is used by the event handler to retrieve the events as JSON.
+  rpc GetUnstructuredEvents(GetUnstructuredEventsRequest) returns (EventsUnstructured);
```

Comment:
Have you considered instead of adding another set of "unstructured" counterparts for everything, just adding "unstructured" or "raw" flag to existing GetEventsRequest and then just adding "UnstructuredEvent" (or "RawEvent" or "JSONEvent") to the events.OneOf?